spock-core/src/main/java/org/spockframework/mock/runtime/IMockMaker.java

```diff
@@ -0,0 +1,131 @@
+package org.spockframework.mock.runtime;
+
+import org.spockframework.mock.CannotCreateMockException;
+import org.spockframework.mock.IMockObject;
+import org.spockframework.util.Beta;
+import org.spockframework.util.Immutable;
+import org.spockframework.util.Nullable;
+import org.spockframework.util.ThreadSafe;
+import spock.mock.MockMakerId;
+
+import java.util.List;
+
+/**
+ * The {@link IMockMaker} provides a {@link java.util.ServiceLoader} extension point to allow the usage
+ * of different implementations to create the underlying mock classes and mock instances.
+ *
+ * <p>
+ * The class {@link MockMakerId} lists the well known {@link IMockMaker} currently defined in Spock.
+ *
+ * <p> You can override the use {@link IMockMaker} with the system property {@value #MOCK_MAKER_PROPERTY} set
+ * to the ID of the {@link IMockMaker}. Please note that the {@link MockMakerId#dynamicProxy} has
+ * always priority over every other {@link IMockMaker}. </p>
+ *
+ * <p>Note: The implementation must be thread-safe, because it can be called concurrently from different threads.</p>
+ */
+@Beta
+@ThreadSafe
+public interface IMockMaker {
+  /**
+   * The system property, which allows to override the priority of the {@link IMockMaker}.
+   *
+   * <p>Please note that the {@link MockMakerId#dynamicProxy} has
+   * always priority over every other {@link IMockMaker}.
+   */
+  String MOCK_MAKER_PROPERTY = "org.spockframework.mock.runtime.mockmaker";
+
+  /**
+   * Returns the ID of this {@link IMockMaker}.
+   * The ID must be unique and should be in the form of {@code your-name-mock-maker}.
+   *
+   * @return the ID
+   */
+  String getId();
+
+  /**
+   * @return The priority in which the {@link IMockMaker}s are applied.
+   */
+  int getPriority();
+
+  /**
+   * Creates a new mock/stub/spy instance for the passed {@link IMockCreationSettings}
+   *
+   * @param settings the settings
+   * @return the new mock instance
+   * @throws CannotCreateMockException if this {@link IMockMaker} can't create such a mock.
+   */
+  Object makeMock(IMockCreationSettings settings) throws CannotCreateMockException;
+
+  ITypeMockability isTypeMockable(IMockCreationSettings settings);
```

Comment:
This is missing JavaDoc.

🤔 Doesn't this check more than a single type, so maybe `isMockable` and `IMockabilityResult` fits better.

Reply:
I have changed the naming and added JavaDoc.